File: interview questions/validateBST.py
import math


class BST:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def validateBst(tree):
    return validateHelper(tree, -math.inf, math.inf)


def validateHelper(tree, min, max):
    if tree == None:
        return True
    if (tree.value < min or tree.value >= max):
        return False
    leftValid = validateHelper(tree.left, min, tree.value)
    rightValid = validateHelper(tree.right, tree.value, max)
    return leftValid and rightValid

File: interview questions/validateBST.js
class BST {
  constructor(value) {
    this.value = value;
    this.left = null;
    this.right = null;
  }
}

function validateBst(tree) {
  return validateHelper(tree, -Infinity, Infinity);
}

function validateHelper(tree, min, max) {
  if (tree === null) return true;
  if (tree.value >= max || tree.value < min) return false;
  const leftValid = validateHelper(tree.left, min, tree.value);
  const rightValid = validateHelper(tree.right, tree.value, max);
  return leftValid && rightValid;
}
